### PR TITLE
Install dependencies of upgrade tasks

### LINKED ISSUE #2643: The `beakerlib` package is missing for the `upgrade` step
When adding an upgrade step which uses beakerlib framework, it seems beakerlib is not being installed and the upgrade tasks fail with `./task.sh: line 2: /usr/share/beakerlib/beakerlib.sh: No such file or directory`

In the upgrade tasks main.fmf:
```
test: ./task.sh
framework: beakerlib
```

Example report (login-logout-stress is not using beakerlib framework):
Screenshot from 2024-01-25 13-50-12

### PR DESCRIPTION
**Pull request by martinhoyer**
Fixes #2643.

Pull Request Checklist

* [x] implement the feature
* [x] extend the test coverage
* [x] include a release note

### CODE REVIEW

**Comment by psss** [comment]
@martinhoyer, could you please include a simple test checking that `beakerlib` would be installed even if the test which is executed is using `shell` as the `framework`? I'd suggest to create a slightly modified copy of `/tests/execute/upgrade/simple` which would select only the shell test:

https://github.com/teemtee/tmt/blob/159b37cd81ef38bc89fd943476a0b263a294d2e5/tests/execute/upgrade/simple.sh#L13-L18

**Review by lukaszachy — COMMENTED** [review]
LGTM.

IMO also worth of release note

**Comment by martinhoyer** [comment]
> Just a note that 'upgrade path' doesn't need to run distro update - it can be any plan. E.g. in [#2643 (comment)](https://github.com/teemtee/tmt/issues/2643#issuecomment-1959707863) it is quite fast. We have 1 existing test which does the distro upgrade and that is IMO enough. All we need to do for this PR is to assert 'upgrade path' dependenicies are installed. No need to waste time in ugrading distro

Thanks @lukaszachy , I'm trying to adapt your reproducer, but I'm not super sure how to modify it so it passes the tmt lint.

btw, I've found a small `dummy-test-package-crested` package to add as `require` in the upgrade task: Total download size: 11 k; Installed size: 6.9 k

**Comment by lukaszachy** [comment]
> I'm trying to adapt your reproducer, but I'm not super sure how to modify it so it passes the tmt lint.

It smells with linter bug. I have same 'warn' reported for in plans for `/tests/execute/upgrade` (in /data).
Especially the `warn C000 value of "how" is not "tmt"` check...

**Comment by psss** [comment]
> IMO also worth of release note

Yeah, makes sense, added one in 376efe07 with some minor adjustments.

File: tmt/steps/execute/upgrade.py [hunk 1]
```diff
@@ -312,8 +335,39 @@ def _perform_upgrade(
                 self._run_discover_upgrade()
                 # Pass in the path-specific env variables
                 extra_environment = plan.environment
+
+            dependencies: list[tmt.base.DependencySimple] = []
+            dependencies_recommend: list[tmt.base.DependencySimple] = []
```
Comment (psss) [inline tmt/steps/execute/upgrade.py:340]: What about calling these two `required_packages` and `recommended_packages` to make it clear at the first sight?

File: tmt/steps/execute/upgrade.py [hunk 2]
```diff
@@ -258,6 +259,28 @@ def _run_discover_upgrade(self) -> None:
...
+            guest: tmt.steps.provision.Guest,
+            dependencies: list[tmt.base.DependencySimple],
+            recommends: bool = False) -> None:
+        """ Install packages required/recommended for upgrade """
+
+        data: _RawPrepareStepData = {
+            'how': 'install',
+            'name': 'recommends' if recommends else 'requires',
+            'summary': 'Install required packages',
+            'order': tmt.utils.DEFAULT_PLUGIN_ORDER_REQUIRES,
```
[the remark this answers is not in the document]
Comment (martinhoyer) [reply tmt/steps/execute/upgrade.py]: removed the line

File: tmt/steps/execute/upgrade.py [hunk 3]
```diff
@@ -258,6 +259,28 @@ def _run_discover_upgrade(self) -> None:
...
+            self,
+            guest: tmt.steps.provision.Guest,
+            dependencies: list[tmt.base.DependencySimple],
+            recommends: bool = False) -> None:
+        """ Install packages required/recommended for upgrade """
+
+        data: _RawPrepareStepData = {
+            'how': 'install',
+            'name': 'recommends' if recommends else 'requires',
+            'summary': 'Install required packages',
```
[the remark this answers is not in the document]
Comment (martinhoyer) [reply tmt/steps/execute/upgrade.py]: are you ok with 'recommends' / 'requires', or do you want to create 'recommended', 'required'?

File: tmt/steps/execute/upgrade.py (hunk 1 above)
Comment (martinhoyer) [reply tmt/steps/execute/upgrade.py]: renamed

Comment (psss) [reply tmt/steps/execute/upgrade.py]: Thanks.

File: tmt/steps/execute/upgrade.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (psss) [reply tmt/steps/execute/upgrade.py]: Let's go with `required-packages-upgrade` as the `name` and `Install packages required by the upgrade` as the `summary` of the install phase.

[the remark this answers is not in the document]
Comment (psss) [reply tmt/steps/execute/upgrade.py]: Renamed in 89c98f5b.

**Review by psss — APPROVED** [review]

**Comment by mmacura311** [comment]
sanity tested this functionality with following copr builds:
```
tmt-1.33.dev29+g2879190a-main.fc40.noarch
tmt+export-polarion-1.33.dev29+g2879190a-main.fc40.noarch
tmt+report-junit-1.33.dev29+g2879190a-main.fc40.noarch
tmt+report-polarion-1.33.dev29+g2879190a-main.fc40.noarch
tmt+provision-beaker-1.33.dev29+g2879190a-main.fc40.noarch
tmt+provision-container-1.33.dev29+g2879190a-main.fc40.noarch
tmt+provision-virtual-1.33.dev29+g2879190a-main.fc40.noarch
tmt+test-convert-1.33.dev29+g2879190a-main.fc40.noarch
tmt+all-1.33.dev29+g2879190a-main.fc40.noarch
```
did run an upgrade test that require beakerlib in the pre-upgrade test already at the beginning, and second one where requirement is added only at the upgrade phase. Both tests passed.

**Pull request merged**